File: Calculate Ninth Digit/R/gen_cusip_checkdigit.R
# Generate Check Digit 
# Purpose: Calculate check digit for 8 digit CUSIPs and return 9 digit CUSIPs: 
# Input: 
#     CUSIP8: Column in a dataframe that contains the CUSIP-8 codes 
#
# Output:
#     Returns a dataframe with a new column, CUSIP9, that is the original CUSIP8 
#     column with a check-digit appended at the end
#
# Contact: BRDS (brds@hbs.edu)
# Repo: https://github.com/hbs-brds/cusip_digits
# Website: https://www.library.hbs.edu/baker-research-data-services


# Define Function 
gen_cusip_checkdigit <- function(CUSIP8) { 

  #Loop through each row of the dataframe
  CUSIP9 <- NA
  for (i in 1:length(CUSIP8)) {
    ## Extract row of dataframe
    x <- CUSIP8[i]
    ## Split string into individual characters
    xlist <- as.list(strsplit(x, "")[[1]])
    
    
    ## Convert special characters 
    xlist2 <- gsub("\\*","36", xlist)
    xlist2 <- gsub("@","37", xlist2)
    xlist2 <- gsub("#","38", xlist2) 
    
    ## Convert letters to numeric 
    xlist2 <- as.numeric(ifelse(grepl("[A-Z]", xlist2), match(xlist2, LETTERS) + 9, xlist2))
    
    
    ## Double every other number
    xlist3 <- xlist2
    xlist3[2] <- xlist3[2]*2
    xlist3[4] <- xlist3[4]*2
    xlist3[6] <- xlist3[6]*2
    xlist3[8] <- xlist3[8]*2
    xlist4 <- xlist3
    
    
    ## Loop through each element in the CUSIP
    for (j in 1:8) {
      ## Sum digits for characters greater than 10
      if (xlist3[j] > 9) {
        
        xlist4[j] = floor(xlist3[j]/10) + (xlist3[j] %% 10)
      } 
    }
    
    
    ## Sum values and calculate check digit 
    xlistsum <- sum(xlist4)
    checkdigit <- floor((10 - (xlistsum %% 10)) %% 10)
    
    
    ## Add final check digit to original CUSIP. 
    CUSIP9[i] <- paste0(x,checkdigit)  
  }  
    ## Return 9 digit CUSIPs
    return(CUSIP9)
  
}

File: Calculate Ninth Digit/Python/gen_cusip_checkdigit.py
"""
Generate Check Digit 
Purpose: Calculate check digit for 8 digit CUSIPs and return 9 digit CUSIPs:  
Input: 
    cusips: a dataframe that contains the relevant CUSIP8 column
Output:
    Returns a dataframe with a new column, CUSIP9, that is the original CUSIP8 
    column with a check-digit appended at the end

Acknowledgments: Thank you Sage Belz for providing us with this code


Repo: https://github.com/hbs-brds/cusip_digits/tree/main
Contact: BRDS (brds@hbs.edu)
"""
import pandas as pd 

# Define function to generate check digit (returns check digit) 
def gen_cusip_checkdigit(cusip):
    val_sum = 0
    digits = cusip.upper()

    # Convert non-numeric digits to values according to their ordinal position in the alphabet plus 9
    # (A=10, B=11,...Z=35) and converting the characters *=36, @=37, #=38.
    alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ*@#"
    for i in range(len(digits)):
        if digits[i].isdigit():
            val = int(digits[i])
        else:
            val = alphabet.index(digits[i]) + 10

        # Multiply every even digit by 2
        if (i % 2) != 0:
            val *= 2

        # If the result of the multiplication is a two-digit number, add the digits together. (12 = 1 + 2 = 3)
        val = (val % 10) + (val // 10)

        # Get the sum of all values.
        val_sum += val

    # Get the floored value of this operation: (10 - (sum modulo 10)) modulo 10.
    check = (10 - (val_sum % 10)) % 10
    return cusip + str(check)


